atomic_reactor/cachito_util.py

```diff
@@ -0,0 +1,164 @@
+        :param burst_retry: int, seconds to wait between retries prior to exceeding
+                            the burst length
+        :param burst_length: int, seconds to switch to slower retry period
+        :param slow_retry: int, seconds to wait between retries after exceeding
+                           the burst length
+        :param timeout: int, when to give up waiting for compose request
+
+        :return: dict, latest representation of the Cachito request
+        :raise CachitoAPIUnsuccessfulRequest: if the request completes unsuccessfully
+        :raise CachitpAPIRequestTimeout: if the request does not complete timely
+        """
+        url = '{}/api/v1/requests/{}'.format(self.api_url, request_id)
+        logger.info('Waiting for request %s to complete...', request_id)
+
+        start_time = time.time()
+        while True:
+            response = self.session.get(url)
+            response.raise_for_status()
+            response_json = response.json()
+
+            state = response_json['state']
+            if state in ('stale', 'failed'):
+                state_reason = response_json.get('state_reason') or 'Unknown'
+                logger.error(dedent("""\
+                   Request %s is in "%s" state: %s
+                   Details: %s
+                   """), request_id, state, state_reason, json.dumps(response_json, indent=4))
+                raise CachitoAPIUnsuccessfulRequest(
+                   'Request {} is in "{}" state: {}'.format(request_id, state, state_reason))
+
+            if state == 'complete':
+                logger.debug('Request %s is complete', request_id)
+                return response_json
+
+            # All other states are expected to be transient and are not checked.
+
+            elapsed = time.time() - start_time
+            if elapsed > timeout:
+                raise CachitpAPIRequestTimeout(
+                    'Request %s not completed after %s seconds' % (url, timeout))
+            else:
+                if elapsed > burst_length:
+                    time.sleep(slow_retry)
+                else:
+                    time.sleep(burst_retry)
+
+    def download_sources(self, request_id, dest_path):
+        """Download the sources from a Cachito request
+
+        :param request_id: int, the Cachito request ID
+        :param dest_path: str, the path to save the sources
+        """
+        logger.debug('Downloading sources bundle from request %d to %s', request_id, dest_path)
+        url = '{}/api/v1/requests/{}/download'.format(self.api_url, request_id)
+        request = self.session.get(url, stream=True)
+        request.raise_for_status()
+
+        with open(dest_path, 'wb') as f:
+            for chunk in request.iter_content(chunk_size=DEFAULT_DOWNLOAD_BLOCK_SIZE):
+                f.write(chunk)
+        logger.debug('Sources bundle for request %d downloaded successfully', request_id)
+
+
+if __name__ == '__main__':
```

Comment:
I can remove this if deemed to be clutter. I found it useful while developing this module.